api/api.go

```diff
@@ -6,40 +6,155 @@ import (
+	api.get(
+		"/topics/{id}",
+		api.isAuthenticated(
+			api.isAuthorised(readPermission, api.getTopicPrivateHandler)), //...ask Eleanor if its this line that should be used ?
+		/*api.isAuthorisedForTopics(readPermission, api.getTopicPrivateHandler*/
+	)
+}
+
+// isAuthenticated wraps a http handler func in another http handler func that checks the caller is authenticated to
+// perform the requested action. handler is the http.HandlerFunc to wrap in an
+// authentication check. The wrapped handler is only called if the caller is authenticated
+func (api *API) isAuthenticated(handler http.HandlerFunc) http.HandlerFunc {
+	return dphandlers.CheckIdentity(handler)
+}
+
+// isAuthorised wraps a http.HandlerFunc another http.HandlerFunc that checks the caller is authorised to perform the
+// requested action. required is the permissions required to perform the action, handler is the http.HandlerFunc to
+// apply the check to. The wrapped handler is only called if the caller has the required permissions.
+func (api *API) isAuthorised(required auth.Permissions, handler http.HandlerFunc) http.HandlerFunc {
+	return api.permissions.Require(required, handler)
+}
+
+// isAuthorised wraps a http.HandlerFunc another http.HandlerFunc that checks the caller is authorised to perform the
+// requested topics action. This authorisation check is specific to topics. required is the permissions required to
+// perform the action, handler is the http.HandlerFunc to apply the check to. The wrapped handler is only called if the
+// caller has the required topic permissions.
+func (api *API) isAuthorisedForTopics(required auth.Permissions, handler http.HandlerFunc) http.HandlerFunc {
+	return api.topicPermissions.Require(required, handler)
+}
+
+// get register a GET http.HandlerFunc.
+func (api *API) get(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("GET")
+}
+
+// get register a PUT http.HandlerFunc.
+func (api *API) put(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("PUT")
+}
+
+// get register a POST http.HandlerFunc.
+func (api *API) post(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("POST")
+}
+
+// get register a DELETE http.HandlerFunc.
+func (api *API) delete(path string, handler http.HandlerFunc) {
+	api.Router.HandleFunc(path, handler).Methods("DELETE")
+}
+
+func (api *API) authenticate(r *http.Request, logData log.Data) bool {
+	var authenticated bool
+
+	if api.enablePrivateEndpoints {
+		var hasCallerIdentity, hasUserIdentity bool
+
+		// NOTE:
+		// If the identity exists then the user has been authenticated.
+		// There is an earlier step in the middleware which will call off to zebedee to
+		// authenticate the request (user/service) and this will add the identity to the
+		// request context for later use in the application ...
+		// ... which happens to be here:
```

Comment:
I don't think this is needed. Personal opinion comments in code are not great there is a misconception about the "value" they offer. In my experience comments go out of dat,  become irrelevant or end up being misleading adding contradictory explanations about code that are no longer true. 

For example how many times have you read function documentation and found it was out of date, missing parameters, declaring exceptions that the function no longer throws?

Reply:
This is not personal opinion it is a statement made in answering many question i put to Nathan about AuthZ and AuthN differences and it does describe what is being done.
Zebedee is getting the info that is put into: caller_identity:Weyland-Yutani Corporation
I did not know how things were being done and it is ALWAYS good to have comments that describe either the how or the why.